chore: update subscriptions docs to include multipart over HTTP (#10663)
apollographql/apollo-client: docs/source/data/subscriptions.mdx

> **Note**: Subscriptions cannot be used to listen to local client events, like subscribing to changes in the cache. Subscriptions are intended to be used to subscribe to external data changes, and have those received changes be stored in the cache. You can then leverage Apollo Client's observability model to watch for changes in the cache, using [`client.watchQuery`](../api/core/ApolloClient#ApolloClient.watchQuery) or [`useQuery`](../api/react/hooks#usequery).

## Supported subscription protocols

The GraphQL spec does not define a specific protocol for sending subscription requests. Apollo Client supports the following protocols for subscriptions:

- WebSocket, using one of the following subprotocols:
    - [`graphql-ws`](https://github.com/enisdenjo/graphql-ws)
    - [`subscriptions-transport-ws`](https://github.com/apollographql/subscriptions-transport-ws) (⚠️ unmaintained)
- HTTP, using [chunked multipart responses](https://github.com/graphql/graphql-over-http/blob/main/rfcs/IncrementalDelivery.md) **(Apollo Client `3.7.11` and later)**

You must use the same protocol as the GraphQL endpoint you're communicating with.

### WebSocket subprotocols

> The first popular JavaScript library to implement subscriptions over WebSocket is called `subscriptions-transport-ws`. **This library is no longer actively maintained.** Its successor is a library called `graphql-ws`. The two libraries _do not use the same WebSocket subprotocol_, so you need to make sure that your server and clients all use the same library.

Reviewer: ```suggestion
The first popular JavaScript library to implement subscriptions over WebSocket is called `subscriptions-transport-ws`. **This library is no longer actively maintained.** Its successor is a library called `graphql-ws`. These two libraries _do not use the same WebSocket subprotocol_, so you need to use the same subprotocol that your GraphQL endpoint uses.
```

> Apollo Client supports both `graphql-ws` and `subscriptions-transport-ws`. We recommend you use the newer library `graphql-ws`, and this page shows how to use it. If you need to use `subscriptions-transport-ws` because your server still uses that protocol, the differences are described [at the bottom of this page](#the-older-subscriptions-transport-ws-library).

Reviewer: ```suggestion
The [WebSocket setup](#websocket-setup) section below uses `graphql-ws`. If your endpoint uses `subscriptions-transport-ws`, see [this section](#the-older-subscriptions-transport-ws-library) for differences in configuration.
```

> **Note**: Confusingly, the `subscriptions-transport-ws` library calls its _WebSocket subprotocol_ `graphql-ws`, and the `graphql-ws` _library_ calls its subprotocol `graphql-transport-ws`! In this article, we refer to the two _libraries_ (`subscriptions-transport-ws` and `graphql-ws`), _not_ the two subprotocols.

### HTTP

> To use Apollo Client with a backend that supports multipart subscriptions over HTTP, no additional client configuration is required. Apollo Client automatically sends the required headers with the request if the terminating `HTTPLink` is passed a subscription operation.

Reviewer: ```suggestion
To use Apollo Client with a GraphQL endpoint that supports [multipart subscriptions over HTTP](https://github.com/graphql/graphql-over-http/blob/main/rfcs/IncrementalDelivery.md), make sure you're using version `3.7.11` or later.

Aside from updating your client version, no additional configuration is required! Apollo Client automatically sends the required headers with the request if the terminating `HTTPLink` is passed a subscription operation.
```

> ## Choosing a subscription library

Reviewer: ```suggestion
## Supported subscription protocols
```

> The GraphQL spec does not define a specific protocol for sending subscription requests. In Apollo Client, subscriptions can be sent over HTTP via multipart transport with the default terminating link, `HTTPLink`, or via WebSocket.

Reviewer: ```suggestion
The GraphQL spec does not define a specific protocol for sending subscription requests. Apollo Client supports the following protocols for subscriptions:

- WebSocket, using one of the following subprotocols:
    - [`graphql-ws`](https://github.com/enisdenjo/graphql-ws)
    - [`subscriptions-transport-ws`](https://github.com/apollographql/subscriptions-transport-ws) (⚠️ unmaintained)
- HTTP, using [chunked multipart responses](https://github.com/graphql/graphql-over-http/blob/main/rfcs/IncrementalDelivery.md) **(Apollo Client x.y.z and later)**

You must use the same protocol as the GraphQL endpoint you're communicating with.
```

> ### Multipart Subscriptions Over HTTP

Author: > Multipart Subscriptions

Review for naming conventions/consistency

> ### Multipart Subscriptions Over HTTP

Reviewer: ```suggestion
### HTTP
```

Reviewer: Recommend keeping it short and expanding in the body text, and also moving this subsection _below_ the WebSocket subsection

To use Apollo Client with a backend that supports multipart subscriptions over HTTP, no additional client configuration is required. Apollo Client automatically sends the required headers with the request if the terminating `HTTPLink` is passed a subscription operation.

> ### WebSocket

Reviewer: ```suggestion
### WebSocket subprotocols
```

> In order to use Apollo Client with a back-end that supports multipart subscriptions over HTTP (for example, Apollo Router) no additional client configuration is required. Apollo Client will automatically send the required headers with the request if the terminating `HTTPLink` receives a subscription operation.

Author: Should we list/link to other docs here besides/in addition to the router?

> In order to use Apollo Client with a back-end that supports multipart subscriptions over HTTP (for example, Apollo Router) no additional client configuration is required. Apollo Client will automatically send the required headers with the request if the terminating `HTTPLink` receives a subscription operation.

Reviewer: ```suggestion
To use Apollo Client with a backend that supports multipart subscriptions over HTTP, no additional client configuration is required. Apollo Client automatically sends the required headers with the request if the terminating `HTTPLink` is passed a subscription operation.
```

> ### WebSocket

Reviewer: ```suggestion
### WebSocket subprotocols
```

## Setting up the transport

> Additional configuration is required if the back-end service your application uses communicates over the WebSocket protocol; in this case, Apollo Client subscriptions most commonly use the [`graphql-ws`](https://www.npmjs.com/package/graphql-ws) library. If the back-end service supports multipart subscriptions via HTTP, however, _no additional configuration is required_.

Author: ```suggestion
Additional configuration is required if the back-end service your application communicates with is using the WebSocket protocol for subscriptions; in this case, Apollo Client most commonly use the [`graphql-ws`](https://www.npmjs.com/package/graphql-ws) library to provide WebSocket support. If the back-end service supports subscriptions over multipart HTTP, however, _no additional configuration is required_.
```

> > As mentioned in [Choosing a subscription library](#choosing-a-subscription-library), some servers that communicate via WebSocket use an older library called `subscriptions-transport-ws`. For necessary changes to use that library, see [below](#the-older-subscriptions-transport-ws-library).

Reviewer: ```suggestion
> As mentioned in [Supported subscription protocols](#supported-subscription-protocols), some GraphQL servers use an older WebSocket library called `subscriptions-transport-ws`. For necessary changes to use that library, see [below](#the-older-subscriptions-transport-ws-library).
```

> ## Subscriptions via Multipart HTTP

Reviewer 2: Some of this feels a bit redundant to the "Multipart Subscriptions Over HTTP" section (same with the websockets). Would it make sense to try and combine these sections together?

Reviewer: Yeah @alessbell I recommend cutting the "Setting up the transport" header, all the way down to line 104 below. And put the callout about `subscriptions-transport-ws`  just below the `WebSocket setup` header

Author: Thanks for the feedback here! Fixed in https://github.com/apollographql/apollo-client/pull/10663/commits/0e7cfbc87694b60b117ac2608b65a88592fa6457.

> No additional libraries or configuration are required. Apollo Client add the required headers to your request when the terminating (and default) `HTTPLink` receives a subscription operation at the `uri` specified when initializing the link or Apollo Client instance.

Reviewer 2: ```suggestion
No additional libraries or configuration are required. Apollo Client adds the required headers to your request when the terminating (and default) `HTTPLink` receives a subscription operation at the `uri` specified when initializing the link or Apollo Client instance.
```

> No additional libraries or configuration are required. Apollo Client add the required headers to your request when the terminating (and default) `HTTPLink` receives a subscription operation at the `uri` specified when initializing the link or Apollo Client instance.

Reviewer 2: ```suggestion
No additional libraries or configuration are required. Apollo Client add the required headers to your request when the default terminating `HTTPLink` receives a subscription operation at the `uri` specified when initializing the link or Apollo Client instance.
```

I think removing the parens and moving "default" before "terminating" makes it read a bit better. Let me know what you think.

Author: I like it, thanks! Applying it manually since it no longer will allow me to commit it here on an outdated diff 👍

> ## Subscriptions via WebSocket

Reviewer: ```suggestion
## WebSocket setup
```
